*** sources/ejabberd_2.1.x/src/odbc/odbc_queries.erl	2011-11-23 01:37:21.000000000 +0100
--- sources/ejabberd_mongo/src/odbc/odbc_queries.erl	2011-11-23 01:37:38.000000000 +0100
***************
*** 5,11 ****
  %%% Created : by Mickael Remond <mremond@process-one.net>
  %%%
  %%%
! %%% ejabberd, Copyright (C) 2002-2011   ProcessOne
  %%%
  %%% This program is free software; you can redistribute it and/or
  %%% modify it under the terms of the GNU General Public License as
--- 5,11 ----
  %%% Created : by Mickael Remond <mremond@process-one.net>
  %%%
  %%%
! %%% ejabberd, Copyright (C) 2002-2010   ProcessOne
  %%%
  %%% This program is free software; you can redistribute it and/or
  %%% modify it under the terms of the GNU General Public License as
***************
*** 90,97 ****
  -define(generic, true).
  -endif.
  
- -include("ejabberd.hrl").
- 
  %% Almost a copy of string:join/2.
  %% We use this version because string:join/2 is relatively
  %% new function (introduced in R12B-0).
--- 90,95 ----
***************
*** 232,237 ****
--- 230,239 ----
  
  users_number(LServer) ->
      case element(1, ejabberd_config:get_local_option({odbc_server, LServer})) of
+     mysql ->
+ 	ejabberd_odbc:sql_query(
+ 	LServer,
+ 	"select table_rows from information_schema.tables where table_name='users'");
      pgsql ->
  	case ejabberd_config:get_local_option({pgsql_users_number_estimate, LServer}) of
  	true ->
***************
*** 521,529 ****
         "where username='", Username, "' and name='", SName, "';"]).
  
  add_privacy_list(Username, SName) ->
      ejabberd_odbc:sql_query_t(
!       ["insert into privacy_list(username, name) "
!        "values ('", Username, "', '", SName, "');"]).
  
  set_privacy_list(ID, RItems) ->
      ejabberd_odbc:sql_query_t(
--- 523,532 ----
         "where username='", Username, "' and name='", SName, "';"]).
  
  add_privacy_list(Username, SName) ->
+     Id = randoms:get_string(),
      ejabberd_odbc:sql_query_t(
!       ["insert into privacy_list(username, name, id) "
!        "values ('", Username, "', '", SName, "', '", Id, "');"]).
  
  set_privacy_list(ID, RItems) ->
      ejabberd_odbc:sql_query_t(
***************
*** 880,893 ****
        ["select count(*) from ", Table, " with (nolock) ", WhereClause]).
  
  get_roster_version(LServer, LUser) ->
!     ejabberd_odbc:sql_query(
!       LServer,
!       ["EXECUTE dbo.get_roster_version '", LUser, "'"]).
! 
! set_roster_version(Username, Version) ->
!     %% This function doesn't know the vhost, so we hope it's the first one defined:
!     LServer = ?MYNAME,
!     ejabberd_odbc:sql_query(
!       LServer,
!       ["EXECUTE dbo.set_roster_version '", Username, "', '", Version, "'"]).
  -endif.
--- 883,890 ----
        ["select count(*) from ", Table, " with (nolock) ", WhereClause]).
  
  get_roster_version(LServer, LUser) ->
! 	ejabberd_odbc:sql_query(LServer, 
! 		["select version from dbo.roster_version with (nolock) where username = '", LUser, "'"]).
! set_roster_version(LUser, Version) ->
! 	update_t("dbo.roster_version", ["username", "version"], [LUser, Version], ["username = '", LUser, "'"]).
  -endif.
